[sociald] Use server timestamps when delta-syncing google calendar event...

antseppa opened the pull request:
...s

Google calendar sync plugin uses timestamp from previous completed buteo sync
round when requesting events for delta sync. This causes problems if the
device clock is behind the server clock. Say the following takes place:
user creates an event on device, upsyncs it at which point the sync plugin
saves a local timestamp. Then user edits the same event and saves it,
which triggers sync. Sync plugin requests changed events by using the timestamp
from previous round. As that timestamp is behind the server clock, the response
also includes the event user just edited. Sync plugin thinks that both
local and server side versions of the event have changed.
In that case, the server side version wins, effectively canceling the
changes user just did.

This commit takes timestamps from server responses, stores them
per-calendar basis to libsocialcache database and uses those values when
requesting events for delta sync.

chriadam commented on src/google/google-calendars/googlecalendarsyncadaptor.cpp:
```diff
@@ -718,6 +728,17 @@ void GoogleCalendarSyncAdaptor::eventsFinishedHandler()
+                  QString(QLatin1String("Setting updated timestamp for Google account: %1. Calendar Id: %2.  Timestamp: %3."))
+                     .arg(accountId).arg(calendarId).arg(updated));
+        }
         updateLocalCalendarNotebookEvents(accountId, accessToken, calendarId, since);
```
should we use the "since" value here, or the "updated" timestamp value from the server?

antseppa replied:
We should use 'since' because the value is used for building a list of events which have been locally modified since the last buteo sync and event modification timestamp in mkal db is in device time.

chriadam commented:
To trigger clean sync when db is upgrade, can you just call purgeDataForOldAccounts() on every accountId contained in the database when the upgrade is done?  Subsequent syncs using that sync profile (ie, for that account) will then be "new clean syncs" I think.

chriadam commented on src/google/google-calendars/googlecalendarsyncadaptor.cpp:
```diff
@@ -588,37 +589,43 @@ void GoogleCalendarSyncAdaptor::updateLocalCalendarNotebooks(int accountId, cons
-    // be synced correctly... XXX TODO: FIXME!!!
-    QDateTime since = needCleanSync
-                    ? QDateTime()
-                    : lastSyncTimestamp(QLatin1String("google"),
```
the lastSyncTimestamp function, IIRC, looks at a sync anchors database table (which is completely separate to the libsocialcache database).  I wonder if we can actually remove that Sync Anchors database altogether (would save some I/O etc).  I think we need to look carefully at the other sync adaptors to see if I made a similar mistake elsewhere - if so, we either need to: 1) use the sync anchors database instead of removing it, but store the the server-supplied timestamps instead of the local ones, 2) remove the sync anchors database and store the server-supplied timestamps in band with the synced data (eg, in the contacts database, etc).

antseppa commented:
> To trigger clean sync when db is upgrade, can you just call purgeDataForOldAccounts() on every 
> accountId contained in the database when the upgrade is done? Subsequent syncs using that sync 
> profile (ie, for that account) will then be "new clean syncs" I think.

The problem is to know when to do that. When db version is increased, libsocialcache nicely recreates the database but that happens silently and has already happened when GoogleCalendarSyncAdaptor constructor is called and AbstractSocialCacheDatabase::isValid() is again true at that point.

antseppa commented:
Added one more commit. Earlier, if an event was both deleted on server side and modified locally, we ended up to a situation where the modified event was upsynced to server, canceling the delete, but then deleted locally anyway.

chriadam commented:
LGTM

chriadam commented:
Note: this will need to be rebased on top of master, as the TRACE(SOCIALD_ERROR, ...) macro form has been removed and replaced with SOCIALD_ERROR_LOG(...) form, in master.

antseppa commented:
Rebased. To force clean sync after db update, I suppose I can just add a oneshot which turns "was last sync successful" -value to false, which will trigger clean sync.

antseppa commented:
...or, since there is a per-account QSetting file, start to maintain plugin version number in it and force clean sync if it has increased.

antseppa commented:
Added a small commit, which maintains sync plugin version number in existing QSetting file and forces clean sync if that has increased. If that approach is ok then this is ready to go in.

chriadam commented:
Yep, that seems fine.  LGTM.
